course03/project01/balint/application.cc

```diff
@@ -0,0 +1,257 @@
+};
+
+void prepareSettings(BasicApp::Settings* settings)
+{
+  settings->setMultiTouchEnabled(false);
+  settings->setWindowSize(800, 600);
+}
+
+void BasicApp::update()
+{
+  static const double timestep = 1.0 / 60.0;
+
+  // Keep track of time.
+  static double time = getElapsedSeconds();
+  static double accumulator = 0.0;
+
+  // Calculate elapsed time since last frame.
+  double elapsed = getElapsedSeconds() - time;
+  time += elapsed;
+  accumulator += math<double>::min(elapsed, 0.1);  // prevents 'spiral of death'
+  while ((accumulator >= timestep) and not mGameOver)
+    {
+      accumulator -= timestep;
+      moveBall();
+      if (mLeftBat.getCenter().y + 200 < mBallPos.y)
+	{
+	  moveLeftBat(false);
+	}
+      else if (mLeftBat.getCenter().y - 200 > mBallPos.y)
+	{
+	  moveLeftBat(true);
+	}
+    }
+}
+void BasicApp::moveBall()
+{
+  if (ballCollidesDown_p())
+    {
+      // std::cout<<"Ball collides down"<<std::endl;
+      mBallVel.y *= -1;
+    }
+  else if (ballCollidesUp_p())
+    {
+      // std::cout<<"Ball collides top"<<std::endl;
+      mBallVel.y *= -1;
+    }
+  else if (ballCollidesWithRect(mLeftBat))
+    {
+      // std::cout<<"Ball collides left"<<std::endl;
+      mBallVel.x *= -1;
+    }
+  else if (ballCollidesWithRect(mRightBat))
+    {
+      // std::cout<<"Ball collides right"<<std::endl;
+      mBallVel.x *= -1;
+    }
+  else if (ballCollidesWithRect(mRightBarrier))
+    {
+      std::cout << "GAME OVER: YOU LOST" << std::endl;
+      ;
+      mGameOver = true;
+    }
+  else if (ballCollidesWithRect(mLeftBarrier))
+    {
+      std::cout << "GAME OVER: YOU WON" << std::endl;
+      ;
+      mGameOver = true;
+    }
+  mBallPos = advanceBall();
+}
+vec2 BasicApp::advanceBall() { return mBallPos + mBallVel; }
+
+void BasicApp::setup()
+{
+  // init
+
+  float halfHeight = height / 2;
+  float halfWidth = width / 2;
+  float mLeftBatPos = halfHeight;
+  float mRightBatPos = halfHeight;
+  mBallPos = vec2(halfWidth, halfHeight);
+  mBallVel = vec2(1, 1);
+
+  mLeftBarrier = Rectf(0, 0, mBarrierWidth, height);
+  mRightBarrier = Rectf(width - mBarrierWidth, 0, width, height);
+  mTopBarrier = Rectf(0, 0, width, mBarrierWidth);
+  mBottomBarrier = Rectf(0, height - mBarrierWidth, width, height);
+
+  mLeftBat = Rectf(mBarrierWidth, mLeftBatPos - 0.5 * mBatLength, 2 * mBarrierWidth, mLeftBatPos + 0.5 * mBatLength);
+  mRightBat = Rectf(width - 2 * mBarrierWidth, mRightBatPos - 0.5 * mBatLength, width - mBarrierWidth,
+		    mRightBatPos + 0.5 * mBatLength);
+}
+
+void BasicApp::moveLeftBat(bool up_p)
+{
+  vec2 move = vec2(0, up_p ? -1 : 1);
+  // std::cout<<"MOving left bat up: "<<up_p<<std::endl;
+  if ((mLeftBat.getCenter() + move).x > 0 && (mLeftBat.getCenter() + move).x < height)
+    {
+      mLeftBat.offset(move);
+    }
+}
+void BasicApp::moveRightBat(bool up_p)
+{
+  vec2 move = vec2(0, up_p ? -20 : 20);
+  // std::cout<<"MOving right bat up: "<<up_p<<std::endl;
+  mRightBat.offset(move);
```

Comment:
This is a general comment on the design of your game: it would have been much better to write several small, isolated, "pure" C++ classes instead of a big monolithic "Game" class. For example, looking at those two methods `moveLeftBat` and `moveRightBat`: `Bat` should be a class, with a `Move` method. It doesn't matter if it's the left or right bat, the shape, move logic etc are the same. To elaborate on why having small classes are better than everything together (from the strongest to weakest reason)
 - code decoupling / separation of responsibilities: if each of your class has a single responsibility, code will be simpler, easier to understand, will grow in a more organic way. imagine again that in a real world application, this pong game which is (already) 250 lines of code will grow, and feature after feature one day it will be huge single file of 2k lines, and it will be harder and harder to not write bugs or just add features - because of the strong coupling between things
 - unit testing: this is a consequence of my first point. without decoupling  / interfaces, you cant write unit test. I know you didnt write tests for this pong game and that's fine I guess, but no real world application can be written without test
 - dealing with high level concepts such as classes help the next developers reading your code. the pong is quite simple and the calculation of coordinates / dimensions ar all over the place while they should be isolated in one place, commented.
 - code reusability: as I said earlier, by using objects instead of raw points/float/rect, you reuse and factorize code